File: banking-web/src/logic/Account.ts
/**
 * Abstract class representing a general bank account.
 * Ported from Java to TypeScript.
 */
export abstract class Account {
  protected accountNumber: string;
  protected accountHolder: string;
  protected balance: number;

  constructor(accountNumber: string, accountHolder: string, initialBalance: number) {
    this.accountNumber = accountNumber;
    this.accountHolder = accountHolder;
    this.balance = initialBalance;
  }

  public deposit(amount: number): string {
    if (amount > 0) {
      this.balance += amount;
      return `Deposited: ${amount}. New Balance: ${this.balance.toFixed(2)}`;
    } else {
      return "Invalid deposit amount.";
    }
  }

  /**
   * Abstract method for withdrawal.
   */
  public abstract withdraw(amount: number): { success: boolean; message: string };

  public getDetails() {
    return {
      accountNumber: this.accountNumber,
      accountHolder: this.accountHolder,
      balance: this.balance,
      type: this.constructor.name,
    };
  }

  public getAccountNumber(): string {
    return this.accountNumber;
  }

  public getBalance(): number {
    return this.balance;
  }
}

File: Account.java
/**
 * Abstract class representing a general bank account.
 * Demonstrates basic OOP concepts like encapsulation and abstraction.
 */
public abstract class Account {
    protected String accountNumber;
    protected String accountHolder;
    protected double balance;

    public Account(String accountNumber, String accountHolder, double initialBalance) {
        this.accountNumber = accountNumber;
        this.accountHolder = accountHolder;
        this.balance = initialBalance;
    }

    public void deposit(double amount) {
        if (amount > 0) {
            balance += amount;
            System.out.println("Deposited: " + amount + ". New Balance: " + balance);
        } else {
            System.out.println("Invalid deposit amount.");
        }
    }

    /**
     * Abstract method for withdrawal. To be implemented by subclasses.
     * Demonstrates polymorphism.
     */
    public abstract void withdraw(double amount);

    public void displayDetails() {
        System.out.println("Account Number: " + accountNumber);
        System.out.println("Account Holder: " + accountHolder);
        System.out.println("Current Balance: " + balance);
    }

    public String getAccountNumber() {
        return accountNumber;
    }

    public double getBalance() {
        return balance;
    }
}
